pkg/scheduler/metrics/metric_recorder.go

```diff
@@ -128,8 +150,34 @@ func (r *MetricAsyncRecorder) ObserveQueueingHintDurationAsync(pluginName, event
 	r.observeMetricAsync(queueingHintExecutionDuration, value, pluginName, event, hint)
 }
 
+// ObserveInFlightEventsAsync observes the in_flight_events metric.
+// Note that this function is not goroutine-safe;
+// we don't lock the map deliberately for the performance reason and we assume the queue (i.e., the caller) takes lock before updating the in-flight events.
+func (r *MetricAsyncRecorder) ObserveInFlightEventsAsync(eventLabel string, valueToAdd float64) {
+	r.aggregatedInflightEventMetric[gaugeVecMetricKey{metricName: InFlightEvents.Name, labelValue: eventLabel}] += int(valueToAdd)
+
+	// Only flush the metric to the channel if the interval is reached.
```

Comment:
Why not implement this as part of `FlushMetrics`?

In tests, we need to ensure that, at the end of the test, the gauge is zero. In the current implementation, if the last call to `ObserveInFlightEventsAsync` happens when `time.Since(r.aggregatedInflightEventMetricLastFlushTime) > r.interval` is false, then the gauge might never reach zero.

Reply:
> Why not implement this as part of FlushMetrics?

If we move flushing logic to `FlushMetrics`, we'd end up needing a lock for a map `aggregatedInflightEventMetric`, which I want to avoid.

> In tests, we need to ensure that, at the end of the test, the gauge is zero. In the current implementation, if the last call to ObserveInFlightEventsAsync happens when time.Since(r.aggregatedInflightEventMetricLastFlushTime) > r.interval is false, then the gauge might never reach zero.

We already have `len(r.aggregatedInflightEventMetric) != 0` check after the last call to `ObserveInFlightEventsAsync`. Or do you mean something different?

Reply:
Assume we do the last `.Done` for all pods, which will call `ObserveInFlightEventsAsync`, but `time.Since(r.aggregatedInflightEventMetricLastFlushTime) > r.interval` happens to be false.

What will cause the flush to happen?